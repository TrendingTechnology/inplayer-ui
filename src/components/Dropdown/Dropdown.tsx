import React, { SelectHTMLAttributes, ChangeEvent } from 'react';
import DropdownContainer from './DropdownContainer';

type Option = {
  value: string;
  displayName: string;
};

type DefaultOption = {
  displayName: string;
  disabled?: boolean;
};

type Props = SelectHTMLAttributes<HTMLSelectElement> & {
  modifiers?: Array<string>;
  value: string;
  onChange?: (value: string | number) => void;
  color?: string;
  options: Array<Option>;
  disabled?: boolean;
  defaultOption?: DefaultOption;
};

const Dropdown: React.FC<Props> = ({ options, onChange = () => null, defaultOption }) => (
  <DropdownContainer onChange={(e) => onChange(e.target.value)}>
    {defaultOption && (
      <option value="" disabled={defaultOption.disabled}>
        {defaultOption.displayName}
      </option>
    )}
    {options.map(({ value, displayName }) => (
      <option value={value} key={value}>
        {displayName}
      </option>
    ))}
  </DropdownContainer>
);

export default Dropdown;
