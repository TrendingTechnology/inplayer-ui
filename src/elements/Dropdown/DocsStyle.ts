import styled from 'styled-components';

export const DropdownWrapperDocs = styled.div`
  > * {
    display: inline;
    margin: 10px;
  }
`;
