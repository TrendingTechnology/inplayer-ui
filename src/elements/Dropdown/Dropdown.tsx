import React, { SelectHTMLAttributes, CSSProperties, ChangeEvent } from 'react';
import DropdownContainer from './DropdownContainer';

type Option = {
  value: string;
  displayName: string;
};

type DefaultOption = {
  displayName: string;
  disabled?: boolean;
};

type Props = SelectHTMLAttributes<HTMLSelectElement> & {
  modifiers?: Array<string>;
  value: string;
  onChange: (e: ChangeEvent<HTMLSelectElement>) => void;
  color?: string;
  options: Array<Option>;
  disabled?: boolean;
  defaultOption?: DefaultOption;
};

const Dropdown: React.FC<Props> = ({ options, onChange, defaultOption, ...rest }) => (
  <DropdownContainer onChange={(e) => onChange(e)} {...rest}>
    {defaultOption && (
      <option value="" disabled={defaultOption.disabled}>
        {defaultOption.displayName}
      </option>
    )}
    {options.map(({ value, displayName }) => (
      <option value={value} key={value}>
        {displayName}
      </option>
    ))}
  </DropdownContainer>
);

export default Dropdown;
