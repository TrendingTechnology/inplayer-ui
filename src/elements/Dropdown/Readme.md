A simple Dropdown controlled component.
Receives an `options` prop which is an array of `{ value, displayName }` pairs. `value` is passed to the `option` tag as `value` prop while `displayName` is the value used in the dropdown UI.

Can accept a custom `color` parameter, otherwise it uses the theme `primary.main` color.

### Standard dropdown

```jsx
const options = [
  { value: 'Option 1', displayName: 'Option 1' },
  { value: 'Option 2', displayName: 'Option 2' },
];

<Dropdown options={options} value={options[0].value} onChange={console.log} />;
```

### Font size modifiers

```jsx
const options = [
  { value: 'Option 1', displayName: 'Option 1' },
  { value: 'Option 2', displayName: 'Option 2' },
];

<div>
  <Dropdown
    modifiers={['fontSizeExtraSmall']}
    options={options}
    value={options[0].value}
    onChange={console.log}
  />
  <br />
  <br />
  <Dropdown
    modifiers={['fontSizeSmall']}
    options={options}
    value={options[0].value}
    onChange={console.log}
  />
  <br />
  <br />
  <Dropdown
    modifiers={['fontSizeMedium']}
    options={options}
    value={options[0].value}
    onChange={console.log}
  />
  <br />
  <br />
  <Dropdown
    modifiers={['fontSizeLarge']}
    options={options}
    value={options[0].value}
    onChange={console.log}
  />
</div>;
```

### With custom color

```jsx
const options = [
  { value: 'Option 1', displayName: 'Option 1' },
  { value: 'Option 2', displayName: 'Option 2' },
];

<Dropdown
  options={options}
  value={options[0].value}
  onChange={console.log}
  color="#936"
/>;
```

### Disabled

```jsx
const options = [
  { value: 'Option 1', displayName: 'Option 1' },
  { value: 'Option 2', displayName: 'Option 2' },
];

<Dropdown
  options={options}
  value={options[0].value}
  onChange={console.log}
  color="#936"
  disabled
/>;
```
