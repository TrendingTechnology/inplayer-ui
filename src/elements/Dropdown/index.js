// @flow
import React from 'react';
import styled from 'styled-components';
import { prop, ifProp } from 'styled-tools';
import { applyStyleModifiers } from 'styled-components-modifiers';
import { uiColors, fontWeights, fontSizes } from 'utils';
import * as fontSizeModifiers from 'modifiers';
import colors from 'config/colors';

const DropdownContainer = styled.select`
  border: 1px solid ${colors.gray};
  background-color: ${colors.white};
  border-radius: 2px;
  color: ${uiColors('text.light')};
  cursor: pointer;
  font-size: ${fontSizes('small')};
  font-weight: ${fontWeights('light')};
  outline: none;
  padding: 0.2em;
  transition: all 0.3s ease;

  &:hover {
    color: ${ifProp('color', prop('color'), uiColors('primary.main'))};
  }

  &:hover,
  &:focus {
    border-color: ${ifProp('color', prop('color'), uiColors('primary.main'))};
  }

  > option {
    color: ${uiColors('text.light')};
    font-weight: ${fontWeights('light')};
  }

  ${applyStyleModifiers(fontSizeModifiers)};
`;

type Option = {
  value: string,
  displayName: string,
};

type Props = HTMLSelectElement & {
  value: string,
  onChange: (value: string) => any,
  className?: string,
  color?: string,
  /** `Option` is a `{ value: string, displayName: string }` object */
  options: Array<Option>,
  style?: Object,
};

const Dropdown = ({ options, onChange, style, className, ...rest }: Props) => {
  const onDropdownChange = (e: SyntheticEvent<*>) => {
    if (typeof onChange === 'function') {
      onChange(e.target.value);
    }
  };
  return (
    <DropdownContainer onChange={onDropdownChange} className={className} style={style} {...rest}>
      {options.map(({ value, displayName }) => (
        <option value={value} key={value}>
          {displayName}
        </option>
      ))}
    </DropdownContainer>
  );
};

Dropdown.defaultProps = {
  className: '',
  style: {},
};

export default Dropdown;
