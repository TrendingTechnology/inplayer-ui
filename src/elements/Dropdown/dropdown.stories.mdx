import { useState } from 'react';
import { Meta, Story, Preview, Props } from "@storybook/addon-docs/blocks";
import Dropdown from '.'
import { DropdownWrapperDocs } from './DocsStyle';

<Meta title="Dropdown" component={<Dropdown/>} />

# Dropdown

A simple Dropdown controlled component which receives an `options` prop which is an array of `{ value, displayName }` pairs 
and `defaultOption` prop which is an object of `{displayName, disabled}`. 

`value` is passed to the option tag as value prop while `displayName` is the value used in the dropdown UI.

Can accept a custom `color` parameter, otherwise it uses the theme `primary.main` color from the library itself.

```jsx
import { Dropdown } from '@inplayer-org/inplayer-ui';

const Page = () => (
  <Dropdown onChange={(e) => setValue(e.target.value)} options={options} value={value} />
)
```


### Standard dropdown
<Preview withSource="open">
  <Story name="standard">
    { ()=> {
    const options = [
        { value: 'Option 1', displayName: 'Option 1' },
        { value: 'Option 2', displayName: 'Option 2' },
    ];
    const [value, setValue] = useState(options[0].value);
    const defaultOption = { displayName: 'default value', disabled: true };
    return (
        <>
            <p>Selected option: {value}</p>
            <Dropdown onChange={(e) => setValue(e.target.value)} options={options} value={value} />
        </>
    )
  }}
  </Story>
</Preview>

### Standard dropdown with default option enabled
<Preview withSource="open">
  <Story name="default option">
    { ()=> {
    const options = [
        { value: 'Option 1', displayName: 'Option 1' },
        { value: 'Option 2', displayName: 'Option 2' },
    ];
    const [value, setValue] = useState(options[0].value);
    const defaultOption = { displayName: 'default value', disabled: true };
    return (
        <>
            <p>Selected option: {value}</p>
            <Dropdown onChange={(e) => setValue(e.target.value)} defaultOption={defaultOption} options={options} value={value} />
        </>
    )
  }}
  </Story>
</Preview>

### Font size modifiers
<Preview withSource="open">
  <Story name="size modifiers">
    { ()=> {
    const options = [
        { value: 'Option 1', displayName: 'Option 1' },
        { value: 'Option 2', displayName: 'Option 2' },
    ];
    const [fontSizeExtraSmallValue, setFontSizeExtraSmallValue] = useState(options[0].value);
    const [fontSizeSmallValue, setFontSizeSmallValue] = useState(options[0].value);
    const [fontSizeMediumValue, setFontSizeMediumValue] = useState(options[0].value);
    const [fontSizeLargeValue, setFontSizeLargeValue] = useState(options[0].value);
    const defaultOption = { displayName: 'default value', disabled: true };
    return (
        <DropdownWrapperDocs> 
            <Dropdown
                modifiers={['fontSizeExtraSmall']}
                options={options}
                value={fontSizeExtraSmallValue}
                onChange={(e) => setFontSizeExtraSmallValue(e.target.value)}
            />
            <Dropdown
                modifiers={['fontSizeSmall']}
                options={options}
                value={fontSizeSmallValue}
                onChange={(e) => setFontSizeSmallValue(e.target.value)}
            />
            <Dropdown
                modifiers={['fontSizeMedium']}
                options={options}
                value={fontSizeMediumValue}
                onChange={(e) => setFontSizeMediumValue(e.target.value)}
            />
            <Dropdown
                modifiers={['fontSizeLarge']}
                options={options}
                value={fontSizeLargeValue}
                onChange={(e) => setFontSizeLargeValue(e.target.value)}
            />
        </DropdownWrapperDocs>
    )
  }}
  </Story>
</Preview>

### Custom color Dropdown
<Preview withSource="open">
  <Story name="custom color">
    { ()=> {
    const options = [
        { value: 'Option 1', displayName: 'Option 1' },
        { value: 'Option 2', displayName: 'Option 2' },
    ];
    const [value, setValue] = useState(options[0].value);
    const defaultOption = { displayName: 'default value', disabled: true };
    return (
        <>
            <Dropdown
                options={options}
                value={value}
                onChange={(e) => setValue(e.target.value)}
                color="#936"
            />
        </>
    )
  }}
  </Story>
</Preview>

### Disabled Dropdown
<Preview withSource="open">
  <Story name="disabled">
    { ()=> {
    const options = [
        { value: 'Option 1', displayName: 'Option 1' },
        { value: 'Option 2', displayName: 'Option 2' },
    ];
    const [value, setValue] = useState(options[0].value);
    return (
        <>
            <Dropdown
                options={options}
                value={value}
                onChange={(e) => setValue(e.target.value)}
                disabled
            />
        </>
    )
  }}
  </Story>
</Preview>