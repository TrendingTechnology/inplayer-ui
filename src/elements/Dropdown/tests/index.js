import { generateSnapshots } from 'helpers';

import Dropdown from '../index';

generateSnapshots(Dropdown);
