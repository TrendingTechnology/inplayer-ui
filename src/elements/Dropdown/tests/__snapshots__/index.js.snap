// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown renders correctly 1`] = `
.c0 {
  border: 1px solid #d7dde5;
  background-color: #fff;
  border-radius: 2px;
  color: #aaadb4;
  cursor: pointer;
  font-size: 0.875rem;
  font-weight: 300;
  outline: none;
  padding: 0.2em;
  -webkit-transition: all 0.3s ease;
  transition: all 0.3s ease;
}

.c0:hover {
  color: #07AAE6;
}

.c0:hover,
.c0:focus {
  border-color: #07AAE6;
}

.c0 > option {
  color: #aaadb4;
  font-weight: 300;
}

<select
  className="c0"
  onChange={[Function]}
  style={Object {}}
  value="loremIpsum"
>
  <option
    value="loremIpsum"
  >
    Lorem Ipsum
  </option>
</select>
`;
